src/url.js

```diff
@@ -217,9 +220,102 @@ export function getSourceOrigin(url) {
       'Unknown path prefix in url %s', url.href);
   const domainOrHttpsSignal = path[2];
   const origin = domainOrHttpsSignal == 's'
-      ? 'https://' + path[3]
-      : 'http://' + domainOrHttpsSignal;
+      ? 'https://' + decodeURIComponent(path[3])
+      : 'http://' + decodeURIComponent(domainOrHttpsSignal);
   // Sanity test that what we found looks like a domain.
   assert(origin.indexOf('.') > 0, 'Expected a . in origin %s', origin);
   return origin;
 }
+
+/**
+ * Returns the source URL of an AMP document for documents served
+ * on a proxy origin or directly.
+ * @param {string|!Location} url URL of an AMP document.
+ * @return {string}
+ */
+export function getSourceUrl(url) {
+  if (typeof url == 'string') {
+    url = parseUrl(url);
+  }
+
+  // Not a proxy URL - return the URL itself.
+  if (!isProxyOrigin(url)) {
+    return url.href;
+  }
+
+  // A proxy URL.
+  // Example path that is being matched here.
+  // https://cdn.ampproject.org/c/s/www.origin.com/foo/
+  // The /s/ is optional and signals a secure origin.
+  const path = url.pathname.split('/');
+  const prefix = path[1];
+  assert(prefix == 'c' || prefix == 'v',
+      'Unknown path prefix in url %s', url.href);
+  const domainOrHttpsSignal = path[2];
```

Comment:
Ok. I redefined `getSourceOrigin` via `getSourceUrl`. Initially I wanted to avoid an extra parse, but I think that's negligible.